### ISSUE TITLE
[Feature request] Support for "Recursively find git projects and switch between them"?

### ISSUE THREAD

**Issue by macintacos** [liuchengxu/vim-clap#520]
There's an extension on VSCode called [Project Manager](https://marketplace.visualstudio.com/items?itemName=alefragnani.project-manager) that does one thing that I can't seem to find in a vim plugin - the ability to quickly navigate to a git repository somewhere other than the one I'm currently in. This feels like something that would fit really naturally into vim-clap, I just don't have the expertise to know how to do it. Here's a little bit of an explainer.

The way it works is that you point PM at a folder and tell it to recursively "search" for any git repositories. Once they are all indexed, you can just simply invoke a command and switch between them, and it'd all work as if you opened that project in VSCode manually.

This feels like something that could be possible in vim-clap; is it? Ideally it'd be compatible with vim-rooter, but if it's not... I can deal. Any help/direction would be much appreciated.

**Comment by liuchengxu** [comment]
Not familiar with the Project Manager, but each time when you search some direcotry with a monster of huge results using `grep2`, the search results will  be cached, which will be used when you run `grep2` on that direcotry again until the cache is removed(currently it is removed when you reboot the computer). You don't have to switch between them, just specify the direcotry directly when using `grep2`, like `:Clap grep2 ~`.

**Comment by macintacos** [comment]
Yeah unfortunately that's not quite what I mean I don't think. I don't necessarily want to search for any files _within_ the git repositories that I find, I just want to be able to list out all the git repositories in a particular directory, search for a git repo I want to switch to, and then when I hit "enter" all that really happens is that my current vim session `cd`'s to that directory such that running things like `:Clap files` will now be searching a completely different directory. Something like `:Clap projects` but for git repositories to `cd`/switch to.

Probably better described with an example. Right now I have a directory similar to this:

```
~/git-projects/
	↳ some-git-repo/
		↳ .git/
	↳ some-other-git-repo/
		↳ .git/
	↳ not-a-git-repo/
```

In Project Manager, I can tell it to index `git-projects` such that I can quickly just `cd` to `some-git-repo` and `some-other-git-repo` without needing to actually type out `cd ~/git-projects/some-git-repo`. It's "smart" enough to not include `not-a-git-repo` in the results because it's not a git repo.

I hope that makes sense! Please let me know if it does not, I can make a video demo'ing Project Manager, just don't have the time right now.

**Comment by liuchengxu** [comment]
Sounds like you could just implement a new provider `projects` to list the git repos under `~/git-projects`, when you select `some-git-repo`, `cd` to `some-git-repo` in the `sink` of `projects` provider. You can see https://github.com/liuchengxu/vim-clap/blob/master/PROVIDER.md and the existing examples in https://github.com/liuchengxu/vim-clap/tree/master/autoload/clap/provider, should not be hard if you know how to list the repos recursively under `~/git-projects`.

**Comment by macintacos** [comment]
I'm not familiar with vimscript at all but I guess I'll give it a shot 🙂 thanks for the pointers

**Comment by macintacos** [comment]
So I did the next best thing that I could think of since the recursive strategies I was using were really slow. I just used something currently installed on my machine to generate a list of all the directories that I `cd` to constantly on a daily basis (via [zoxide](https://github.com/ajeetdsouza/zoxide)) and used that as a database for a new Clap provider. And it works wonderfully well! Here's the function I created:

```vim
function! GetZoxideDirs()
    let zoxide_dirs = systemlist('zoxide query -l -s')
    let final_dirs = []

    for dir in zoxide_dirs
       let current_dir = "" . dir " adding an icon
       call add(final_dirs, current_dir)
    endfor

    return final_dirs
endfunction

let g:clap_provider_zoxide = {}
let g:clap_provider_zoxide.source = GetZoxideDirs()
let g:clap_provider_zoxide.sink = 'Clap files'
```

Essentially, it'll list all of the directories that have been generated by `zoxide` over time and I can use that as a basis to run `Clap files` on them. It works _really_ well, and I can do custom filtering in zoxide itself and not have to worry about customizing this function any further. After setting this up, I can just run `:Clap zoxide` and it works exactly as I expect.

Thanks for the great tooling!
